File: src/dev_health_ops/api/queries/metrics.py
from __future__ import annotations

from datetime import date
from typing import Any

from dev_health_ops.investment_taxonomy import SUBCATEGORIES, THEMES, theme_of

from .client import query_dicts


def _date_params(start_day: date, end_day: date) -> dict[str, Any]:
    return {"start_day": start_day, "end_day": end_day}


# CHAOS-2377 / CHAOS-2645: daily-rollup tables the sync job re-writes per run.
# work_item_metrics_daily and work_item_user_metrics_daily are
# ReplacingMergeTree(computed_at) (migration 055); work_item_state_durations_daily
# stays MergeTree. In all cases a top-level sum() double-counts re-runs/backfills,
# so the table read is wrapped in a per-key argMax(..., computed_at) dedup subquery
# before aggregating (the FINAL-equivalent for the metric-config read path).
_DEDUP_BY_COMPUTED_AT: dict[str, tuple[str, ...]] = {
    "work_item_state_durations_daily": (
        "day",
        "provider",
        "work_scope_id",
        "team_id",
        "status",
    ),
    "work_item_metrics_daily": (
        "day",
        "provider",
        "work_scope_id",
        "team_id",
    ),
    "work_item_user_metrics_daily": (
        "day",
        "provider",
        "work_scope_id",
        "user_identity",
    ),
}


def _metric_from_clause(
    *,
    table: str,
    column: str,
    scope_filter: str,
    start_param: str = "start_day",
    end_param: str = "end_day",
) -> str:
    """Return the FROM source for a metric read.

    For most tables this is just the raw table name. For ReplacingMergeTree-by-
    ``computed_at`` daily tables (those in ``_DEDUP_BY_COMPUTED_AT``) the table
    is wrapped in a subquery that collapses to the latest ``computed_at`` per
    natural key — so a top-level ``sum(column)`` counts each key once instead of
    once per re-run. ``org_id`` stays filtered in the inner ``WHERE``.

    ``start_param`` / ``end_param`` name the bound query params so a single
    subquery shape can serve both the current and comparison windows (the
    driver-delta CTE pair reuses this with ``compare_start`` / ``compare_end``).
    """
    natural_key = _DEDUP_BY_COMPUTED_AT.get(table)
    if natural_key is None:
        return table
    key_columns = ",\n                ".join(natural_key)
    return f"""(
            SELECT
                {key_columns},
                argMax({column}, computed_at) AS {column}
            FROM {table}
            WHERE day >= %({start_param})s AND day < %({end_param})s
            {scope_filter}
              AND org_id = %(org_id)s
            GROUP BY {", ".join(natural_key)}
        )"""


async def fetch_metric_series(
    client: Any,
    *,
    table: str,
    column: str,
    start_day: date,
    end_day: date,
    scope_filter: str,
    scope_params: dict[str, Any],
    aggregator: str,
    org_id: str = "",
) -> list[dict[str, Any]]:
    value_expression = _metric_value_expression(
        table=table, column=column, aggregator=aggregator
    )
    if table in _DEDUP_BY_COMPUTED_AT:
        # Dedup re-run rows under the outer sum() (CHAOS-2377). The subquery
        # filters day/scope/org_id and emits one row per natural key with the
        # latest computed_at, so we group only by day on the outer query.
        from_clause = _metric_from_clause(
            table=table, column=column, scope_filter=scope_filter
        )
        query = f"""
        SELECT
            day,
            {value_expression} AS value
        FROM {from_clause}
        GROUP BY day
        ORDER BY day
    """
    else:
        query = f"""
        SELECT
            day,
            {value_expression} AS value
        FROM {table}
        WHERE day >= %(start_day)s AND day < %(end_day)s
        {scope_filter}
          AND org_id = %(org_id)s
        GROUP BY day
        ORDER BY day
    """
    params = _date_params(start_day, end_day)
    params.update(scope_params)
    params["org_id"] = org_id
    return await query_dicts(client, query, params)


async def fetch_metric_value(
    client: Any,
    *,
    table: str,
    column: str,
    start_day: date,
    end_day: date,
    scope_filter: str,
    scope_params: dict[str, Any],
    aggregator: str,
    org_id: str = "",
) -> float:
    value_expression = _metric_value_expression(
        table=table, column=column, aggregator=aggregator
    )
    if table in _DEDUP_BY_COMPUTED_AT:
        # Dedup re-run rows under the outer sum() (CHAOS-2377). Without this the
        # /explain blocked_work headline (current + comparison) inflates by the
        # number of duplicate daily runs/backfills for the same natural key.
        from_clause = _metric_from_clause(
            table=table, column=column, scope_filter=scope_filter
        )
        query = f"""
        SELECT
            {value_expression} AS value
        FROM {from_clause}
    """
    else:
        query = f"""
        SELECT
            {value_expression} AS value
        FROM {table}
        WHERE day >= %(start_day)s AND day < %(end_day)s
        {scope_filter}
          AND org_id = %(org_id)s
    """
    params = _date_params(start_day, end_day)
    params.update(scope_params)
    params["org_id"] = org_id
    rows = await query_dicts(client, query, params)
    if not rows:
        return 0.0
    value = rows[0].get("value")
    return float(value or 0.0)


def _metric_value_expression(*, table: str, column: str, aggregator: str) -> str:
    if table == "repo_metrics_daily" and column == "pr_rework_ratio":
        return "SUM(pr_rework_ratio * prs_merged) / NULLIF(SUM(prs_merged), 0)"
    return f"{aggregator}({column})"


async def fetch_blocked_hours(
    client: Any,
    *,
    start_day: date,
    end_day: date,
    scope_filter: str,
    scope_params: dict[str, Any],
    org_id: str = "",
) -> tuple[float, list[dict[str, Any]]]:
    # CHAOS-2377: the daily job appends a fresh row per run to this plain
    # MergeTree, so a re-run/backfill of the same day leaves duplicate rows.
    # Dedup with argMax(duration_hours, computed_at) over the full natural key
    # before summing, matching the operating_review reader; summing raw rows
    # would inflate the blocked-hours panel on every re-run.
    query = f"""
        SELECT
            day,
            sum(duration_hours) AS value
        FROM (
            SELECT
                day,
                provider,
                work_scope_id,
                team_id,
                status,
                argMax(duration_hours, computed_at) AS duration_hours
            FROM work_item_state_durations_daily
            WHERE day >= %(start_day)s AND day < %(end_day)s
              AND status = 'blocked'
            {scope_filter}
              AND org_id = %(org_id)s
            GROUP BY day, provider, work_scope_id, team_id, status
        )
        GROUP BY day
        ORDER BY day
    """
    params = _date_params(start_day, end_day)
    params.update(scope_params)
    params["org_id"] = org_id
    rows = await query_dicts(client, query, params)
    total = sum(float(row.get("value") or 0.0) for row in rows)
    return total, rows


_INVESTMENT_THEME_LABELS = {
    "feature_delivery": "Feature Delivery",
    "operational": "Operational / Support",
    "maintenance": "Maintenance / Tech Debt",
    "quality": "Quality / Reliability",
    "risk": "Risk / Security",
}


def canonical_investment_theme_sql(column: str = "investment_area") -> str:
    mappings: dict[str, str] = {theme: theme for theme in THEMES}
    mappings.update(
        {subcategory: theme_of(subcategory) for subcategory in SUBCATEGORIES}
    )
    for subcategory in SUBCATEGORIES:
        leaf = subcategory.rsplit(".", 1)[-1]
        mappings.setdefault(leaf, theme_of(subcategory))

    clauses = []
    for raw_key, theme in sorted(mappings.items()):
        clauses.append(f"lowerUTF8({column}) = '{raw_key}'")
        clauses.append(f"'{theme}'")
    return f"multiIf({', '.join(clauses)}, '')"


async def fetch_rework_theme_allocation(
    client: Any,
    *,
    start_day: date,
    end_day: date,
    scope_filter: str,
    scope_params: dict[str, Any],
    work_category_filter: str = "",
    work_category_params: dict[str, Any] | None = None,
    org_id: str = "",
) -> list[dict[str, Any]]:
    canonical_theme_expr = canonical_investment_theme_sql("investment_area")
    query = f"""
        SELECT
            canonical_theme AS theme,
            sum(work_items_completed) AS allocation,
            sum(prs_merged) AS prs_merged,
            sum(churn_loc) AS churn_loc
        FROM (
            SELECT
                day,
                repo_id,
                team_id,
                {canonical_theme_expr} AS canonical_theme,
                project_stream,
                argMax(work_items_completed, computed_at) AS work_items_completed,
                argMax(prs_merged, computed_at) AS prs_merged,
                argMax(churn_loc, computed_at) AS churn_loc
            FROM investment_metrics_daily
            WHERE day >= %(start_day)s AND day < %(end_day)s
            {scope_filter}
            {work_category_filter}
              AND org_id = %(org_id)s
            GROUP BY day, repo_id, team_id, canonical_theme, project_stream
        )
        WHERE canonical_theme != ''
        GROUP BY canonical_theme
        ORDER BY allocation DESC
    """
    params = _date_params(start_day, end_day)
    params.update(scope_params)
    params.update(work_category_params or {})
    params["org_id"] = org_id
    rows = await query_dicts(client, query, params)
    rows = [row for row in rows if str(row.get("theme") or "") in THEMES]
    total = sum(float(row.get("allocation") or 0.0) for row in rows)
    allocations: list[dict[str, Any]] = []
    for row in rows:
        theme = str(row.get("theme") or "")
        allocation = float(row.get("allocation") or 0.0)
        allocations.append(
            {
                "theme": theme,
                "label": _INVESTMENT_THEME_LABELS[theme],
                "allocation": allocation,
                "allocation_pct": (allocation / total * 100.0) if total else 0.0,
                "prs_merged": int(row.get("prs_merged") or 0),
                "churn_loc": int(row.get("churn_loc") or 0),
            }
        )
    return allocations

File: cmd/query-api/internal/explain/metrics.go
package explain

import (
	"context"
	"fmt"
	"strings"
	"time"

	dhclickhouse "github.com/full-chaos/dev-health-go/clickhouse"
)

// dedupNaturalKeys ports api/queries/metrics.py's _DEDUP_BY_COMPUTED_AT,
// restricted to the four tables metricConfigs actually reaches (every
// metricConfigs entry's Table is a key of this map -- confirmed against
// migrations 055/096, see this package's own PR RISK-NOTES for the
// per-table citation). The OTHER five tables that dict registers
// (work_item_user_metrics_daily, cicd_metrics_daily,
// incident_metrics_daily, testops_release_confidence,
// testops_pipeline_stability) are unreachable from /explain and are not
// ported here -- a future metric-config entry that adds one of them must
// extend this map too.
var dedupNaturalKeys = map[string][]string{
	"repo_metrics_daily":              {"day", "repo_id"},
	"work_item_state_durations_daily": {"day", "provider", "work_scope_id", "team_id", "status"},
	"work_item_metrics_daily":         {"day", "provider", "work_scope_id", "team_id"},
	"deploy_metrics_daily":            {"day", "repo_id"},
}

// nullableMetricColumns names every metricConfigs column that is
// Nullable(Float64) in its own table's CREATE TABLE (001_metrics_v2.sql):
// work_item_metrics_daily.cycle_time_p50_hours and
// repo_metrics_daily.pr_first_review_p50_hours. The other six columns
// metricConfigs reaches are all non-nullable (UInt32 or a bare Float64,
// confirmed against the same migration), so a plain argMax is correct
// for them.
var nullableMetricColumns = map[string]bool{
	"cycle_time_p50_hours":      true,
	"pr_first_review_p50_hours": true,
}

// metricValueProjection applies the class ruling (b) dedup fix for a
// Nullable metric column: `argMax(col, version)` on a Nullable column can
// return a STALE non-null version instead of the row with the true
// maximal version, if that latest version's own value happens to be
// NULL -- argMax's candidate selection skips a NULL `arg`, it does not
// track "the value at the max version" the way a non-nullable column
// does. The fix is `(argMax(tuple(col), version)).1`: wrapping the
// Nullable column in a tuple makes argMax compare candidates by
// (value IS NOT NULL, value) tuples instead, so the TRUE latest version
// always wins, NULL included. Python's own reader
// (api/queries/metrics.py's _metric_from_clause) is NOT a correct
// exemplar here (bare argMax on a Nullable column) -- this is a
// declared, required divergence under the class ruling, not a
// data-semantics choice.
func metricValueProjection(column string) string {
	if nullableMetricColumns[column] {
		return fmt.Sprintf("(argMax(tuple(%s), computed_at)).1 AS %s", column, column)
	}
	return fmt.Sprintf("argMax(%s, computed_at) AS %s", column, column)
}

// metricFromClause ports _metric_from_clause (api/queries/metrics.py:
// 61-104), restricted to the dedup branch: every table metricConfigs
// names IS in dedupNaturalKeys (see that map's own doc comment), so
// Python's non-dedup "else" branch (a raw `FROM {table}`) is dead code on
// this route's call path and is not ported -- a future metric-config
// entry naming a table outside dedupNaturalKeys would need it added.
//
// ORG SCOPE + class ruling (a): org_id sits inside this SAME subquery's
// WHERE, before the GROUP BY that collapses to one row per natural key --
// never a filter applied after a cross-tenant scan.
//
// DEDUP + class ruling (b): the value projection is
// metricValueProjection's own argMax/tuple-argMax choice, never a bare
// argMax on a Nullable column (see that function's own doc comment).
//
// startParam/endParam name the bound query params so one function serves
// both the current window (start_day/end_day) and the comparison window
// (compare_start/compare_end), matching Python's own start_param/
// end_param default-override parameters.
func metricFromClause(table, column, scopeFilterSQL, startParam, endParam string) string {
	keys := dedupNaturalKeys[table]
	keyColumns := strings.Join(keys, ",\n        ")
	return fmt.Sprintf(`(
    SELECT
        %s,
        %s
    FROM %s
    WHERE day >= {%s:Date} AND day < {%s:Date}
    %s
      AND org_id = {org_id:String}
    GROUP BY %s
)`, keyColumns, metricValueProjection(column), table, startParam, endParam, scopeFilterSQL, keyColumns)
}

// fetchMetricValue ports fetch_metric_value (api/queries/metrics.py:
// 155-198), restricted to the dedup branch (see metricFromClause's own
// doc comment). Returns float(value or 0.0) already applied -- the
// caller (response.go) applies safeFloat on top, matching Python's own
// two-layer None-then-NaN/Inf handling (fetch_metric_value's inner
// `float(value or 0.0)`, then build_explain_response's own outer
// safe_float call).
func (reader *Reader) fetchMetricValue(ctx context.Context, table, column, aggregator string, startDay, endDay time.Time, scopeFilterSQL string, scopeBindings []dhclickhouse.Binding, orgID string) (float64, error) {
	if reader == nil || reader.client == nil {
		return 0, ErrUnavailable
	}

	fromClause := metricFromClause(table, column, scopeFilterSQL, "start_day", "end_day")
	// toFloat64(...) around the outer aggregator: a
	// sum-aggregator metric over an integer column (deploy_freq's
	// deployments_count, churn's total_loc_touched) makes ClickHouse's
	// sum() return UInt64, not Float64 -- confirmed live, this binary's
	// native driver rejects scanning that into the *float64 destination
	// below outright ("converting UInt64 to **float64 is unsupported"),
	// the same class of mismatch newUnrestrictedReadClickHouseOptions'
	// own doc comment (query_route.go) and quadrant.go's own
	// `toFloat64(%s) AS value` fix already document elsewhere in this
	// binary. avg()-aggregated metrics (review_latency,
	// change_failure_rate) already return Float64/Float32 on their own,
	// so this cast is a no-op for them -- never a value change, only a
	// static, driver-safe destination type.
	query := fmt.Sprintf(`
SELECT
    toFloat64(%s(%s)) AS value
FROM %s
%s
`, aggregator, column, fromClause, settingsMaxExecutionTime())

	bindings := append([]dhclickhouse.Binding{
		{Name: "start_day", Value: dateBindingValue(startDay)},
		{Name: "end_day", Value: dateBindingValue(endDay)},
		{Name: "org_id", Value: orgID},
	}, scopeBindings...)

	rows, err := reader.client.Query(ctx, query, bindings)
	if err != nil {
		return 0, fmt.Errorf("fetch metric value: %w", err)
	}
	defer rows.Close()

	if !rows.Next() {
		if err := rows.Err(); err != nil {
			return 0, fmt.Errorf("iterate metric value rows: %w", err)
		}
		return 0.0, nil
	}
	var value *float64
	if err := rows.Scan(&value); err != nil {
		return 0, fmt.Errorf("scan metric value row: %w", err)
	}
	return floatOrZero(value), nil
}

// metricRow is one row fetchMetricContributors/fetchMetricDriverDelta
// returns -- id/value for a contributor row, id/value/delta_pct for a
// driver row (DeltaPct is unused/zero for a contributor row; the caller,
// buildContributor, always overrides a contributor's delta to 0.0
// itself, matching explain.py:240's own `delta_value=0.0` literal).
type metricRow struct {
	ID       string
	Value    float64
	DeltaPct float64
}

// fetchMetricContributors ports fetch_metric_contributors
// (api/queries/explain.py:10-57), dedup branch only (see
// metricFromClause's own doc comment). limit=6 matches explain.py's own
// unqualified call (build_explain_response never overrides the keyword
// default).
//
// aggregator is the metric's OWN config.Aggregator (sum for a count-type
// metric, avg for a ratio/duration-type one) -- the same choice
// fetchMetricValue's headline read already makes. Python's own reader
// hardcodes avg() here regardless of the metric, so a sum-aggregator
// metric's ranking there silently averages a quantity its own headline
// and label present as a total. This is a declared, required divergence,
// not a data-semantics choice.
func (reader *Reader) fetchMetricContributors(ctx context.Context, table, column, groupBy, aggregator string, startDay, endDay time.Time, scopeFilterSQL string, scopeBindings []dhclickhouse.Binding, orgID string) ([]metricRow, error) {
	if reader == nil || reader.client == nil {
		return nil, ErrUnavailable
	}
	const limit = 6

	fromClause := metricFromClause(table, column, scopeFilterSQL, "start_day", "end_day")
	// toFloat64(...) around the outer aggregator: same fix as
	// fetchMetricValue's own doc comment -- a sum-aggregated integer
	// column returns UInt64, which this binary's native driver refuses to
	// scan into a *float64 destination.
	query := fmt.Sprintf(`
SELECT
    toString(%s) AS id,
    toFloat64(%s(%s)) AS value
FROM %s
GROUP BY %s
ORDER BY value DESC
LIMIT {limit:UInt64}
%s
`, groupBy, aggregator, column, fromClause, groupBy, settingsMaxExecutionTime())

	bindings := append([]dhclickhouse.Binding{
		{Name: "start_day", Value: dateBindingValue(startDay)},
		{Name: "end_day", Value: dateBindingValue(endDay)},
		{Name: "org_id", Value: orgID},
		{Name: "limit", Value: limit},
	}, scopeBindings...)

	rows, err := reader.client.Query(ctx, query, bindings)
	if err != nil {
		return nil, fmt.Errorf("fetch metric contributors: %w", err)
	}
	defer rows.Close()

	out := make([]metricRow, 0)
	for rows.Next() {
		var id string
		var value *float64
		if err := rows.Scan(&id, &value); err != nil {
			return nil, fmt.Errorf("scan metric contributor row: %w", err)
		}
		out = append(out, metricRow{ID: id, Value: floatOrZero(value)})
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate metric contributor rows: %w", err)
	}
	return out, nil
}

// fetchMetricDriverDelta ports fetch_metric_driver_delta
// (api/queries/explain.py:60-149), dedup branch only. limit=3 matches
// explain.py's own unqualified call. aggregator is the metric's own
// config.Aggregator -- the same declared divergence from Python's
// hardcoded avg() that fetchMetricContributors' own doc comment states.
// delta_pct is computed IN SQL, the same CASE Python's own query uses (a
// NULL-safe fallback the naive `(current-previous)/previous` would not
// have for previous=0) -- scanned as nullable and denulled the SAME way
// value is, matching where Python's own safe_float(row.get("delta_pct"))
// applies AFTER the SQL layer, not folded into the SQL itself: folding it
// in would change which branch the CASE takes for a NULL current/previous
// pairing (see this package's own PR RISK-NOTES for the worked example
// this avoids).
func (reader *Reader) fetchMetricDriverDelta(ctx context.Context, table, column, groupBy, aggregator string, startDay, endDay, compareStart, compareEnd time.Time, scopeFilterSQL string, scopeBindings []dhclickhouse.Binding, orgID string) ([]metricRow, error) {
	if reader == nil || reader.client == nil {
		return nil, ErrUnavailable
	}
	const limit = 3

	currentFrom := metricFromClause(table, column, scopeFilterSQL, "start_day", "end_day")
	previousFrom := metricFromClause(table, column, scopeFilterSQL, "compare_start", "compare_end")
	// current/previous were a `WITH current AS (...), previous AS (...)`
	// CTE pair until this fix: dev-health-go's client-side read-only
	// guard (clickhouse/client.go's validateReadOnlyStatement) requires a
	// statement's FIRST token to be the literal "SELECT", so a query
	// beginning "WITH ..." is rejected before it ever reaches ClickHouse
	// -- ErrUnsafeStatement ("clickhouse runtime: unsafe statement"),
	// wrapped into the generic 503 every /explain request degraded to,
	// with nothing logged (the defect the telemetry sweep in this same change also
	// fixes). Each CTE is referenced exactly once (current in FROM,
	// previous in the LEFT JOIN), so inlining both as ordinary derived-
	// table subqueries is a purely mechanical, semantically identical
	// rewrite -- no materialization or multiple-reference concern applies.
	// toFloat64(...) around each side's outer aggregator: same this defect class
	// fix as fetchMetricValue's own doc comment -- a sum-aggregated
	// integer column returns UInt64, which this binary's native driver
	// refuses to scan into current.value/previous.value's *float64
	// destinations.
	query := fmt.Sprintf(`
SELECT
    current.id AS id,
    current.value AS value,
    CASE WHEN previous.value = 0 THEN 0 ELSE (current.value - previous.value) / previous.value * 100 END AS delta_pct
FROM (
    SELECT toString(%s) AS id, toFloat64(%s(%s)) AS value
    FROM %s
    GROUP BY %s
) AS current
LEFT JOIN (
    SELECT toString(%s) AS id, toFloat64(%s(%s)) AS value
    FROM %s
    GROUP BY %s
) AS previous ON current.id = previous.id
ORDER BY delta_pct DESC
LIMIT {limit:UInt64}
%s
`, groupBy, aggregator, column, currentFrom, groupBy, groupBy, aggregator, column, previousFrom, groupBy, settingsMaxExecutionTime())

	bindings := append([]dhclickhouse.Binding{
		{Name: "start_day", Value: dateBindingValue(startDay)},
		{Name: "end_day", Value: dateBindingValue(endDay)},
		{Name: "compare_start", Value: dateBindingValue(compareStart)},
		{Name: "compare_end", Value: dateBindingValue(compareEnd)},
		{Name: "org_id", Value: orgID},
		{Name: "limit", Value: limit},
	}, scopeBindings...)

	rows, err := reader.client.Query(ctx, query, bindings)
	if err != nil {
		return nil, fmt.Errorf("fetch metric driver delta: %w", err)
	}
	defer rows.Close()

	out := make([]metricRow, 0)
	for rows.Next() {
		var id string
		var value, deltaPctValue *float64
		if err := rows.Scan(&id, &value, &deltaPctValue); err != nil {
			return nil, fmt.Errorf("scan metric driver row: %w", err)
		}
		out = append(out, metricRow{ID: id, Value: floatOrZero(value), DeltaPct: floatOrZero(deltaPctValue)})
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate metric driver rows: %w", err)
	}
	return out, nil
}

// dateBindingValue formats t as "YYYY-MM-DD" for a {name:Date}-typed
// bound parameter -- NOT a raw time.Time, matching the established fix
// in internal/analytics/validate.go's own dateBindingValue (its doc
// comment: six sites shared a {name:Date}-vs-time.Time formatting defect
// before that fix landed). Duplicated here rather than imported, matching
// this binary's own "repeat, don't couple" convention for a helper this
// narrow.
func dateBindingValue(t time.Time) string {
	year, month, day := t.Date()
	return fmt.Sprintf("%04d-%02d-%02d", year, int(month), day)
}
